### PR TITLE
Analyzer process stderr/out debug info

### PR DESCRIPTION
**Pull request by mattnibs**
When an analyzer process exits with a non-zero code, return an error message
that contains the (truncated) output from stderr and stdout.

Also include a StderrPath and StdoutPath in the analyzer configuration, that
outputs the analyzer process stderr/stdout to the specified path.

Also:

- Add ability to read analyzer configuration from a yaml file. This useful
  functionality was required for tests.

Part of #16

### CODE REVIEW

File: analyzer/config.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+type Config struct {
+	Args       []string       `yaml:"args"`
+	Cmd        string         `yaml:"cmd"`
+	Globs      []string       `yaml:"globs"`
+	Launcher   Launcher       `yaml:"-"`
+	ReaderOpts zio.ReaderOpts `yaml:"-"`
+	Shaper     string         `yaml:"shaper"`
+	StdoutPath string         `yaml:"stdout"`
+	StderrPath string         `yaml:"stderr"`
```
Comment (nwt) [inline analyzer/config.go:18]:
```suggestion
	StdoutPath string         `yaml:"stdout_path"`
	StderrPath string         `yaml:"stderr_path"`
```

File: analyzer/process.go [hunk 2]
```diff
@@ -1,55 +1,201 @@
...
+	Stderr []byte
+	Stdout []byte
 }
 
-func NewProcess(cmd *exec.Cmd) *Process {
-	p := &Process{cmd: cmd, stderrBuf: bytes.NewBuffer(nil)}
-	cmd.Stderr = p.stderrBuf
-	return p
+func (p *ProcessExitError) Error() string {
+	builder := new(strings.Builder)
```
Comment (nwt) [inline analyzer/process.go:114]:
```suggestion
var builder strings.Builder
```

Comment (nwt) [inline analyzer/process.go:114]:
```suggestion
	var builder strings.Builder
```

File: analyzer/process.go [hunk 3]
```diff
@@ -1,55 +1,201 @@
...
+
+	return builder.String()
 }
 
-func (p *Process) Start() error {
-	return wrapError(p.cmd.Start(), p.cmd.Path, p.stderrBuf.String())
+// prefixSuffixSaver is an io.Writer which retains the first N bytes
+// and the last N bytes written to it. The Bytes() methods reconstructs
+// it with a pretty error message.
+// Taken from github.com/golang/go/src/os/exec/exec.go
```
Comment (nwt) [inline analyzer/process.go:138]:
```suggestion
// Taken from github.com/golang/go/blob/master/src/os/exec/exec.go.
```

File: cli/analyzecli/flags.go [hunk 4]
```diff
@@ -19,22 +18,36 @@ var (
...
+		Shaper: suricatashaper,
 	}
 )
 
 type Flags struct {
-	Configs  []analyzer.Config
-	suricata bool
-	zeek     bool
+	configPath     string
+	Configs        []analyzer.Config
```
Comment (nwt) [inline cli/analyzecli/flags.go:27]:
```suggestion
	Configs        []analyzer.Config
	configPath     string
```

File: cli/analyzecli/flags.go [hunk 5]
```diff
@@ -19,22 +18,36 @@ var (
...
+	suricata       bool
+	suricataStderr string
+	suricataStdout string
+	zeek           bool
+	zeekStderr     string
+	zeekStdout     string
 }
 
 func (f *Flags) SetFlags(fs *flag.FlagSet) {
+	fs.StringVar(&f.configPath, "config", "", "path to configuration yaml file")
```
Comment (nwt) [inline cli/analyzecli/flags.go:37]:
```suggestion
	fs.StringVar(&f.configPath, "config", "", "path to YAML configuration file")
```

File: cmd/brimcap/ztests/analyze-process-error.yaml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  export PATH=$(pwd):$PATH
+  chmod +x errorproc.sh
+
+  brimcap analyze -config=config1.yaml alerts.pcap 
+  >&2 echo === 
+  brimcap analyze -config=config2.yaml alerts.pcap
+  ls
+  cat stderr.out
+
+
```
Comment (nwt) [inline cmd/brimcap/ztests/analyze-process-error.yaml:11]:
```suggestion

```

File: cmd/brimcap/ztests/analyze-process-error.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
+script: |
+  export PATH=$(pwd):$PATH
```
Comment (nwt) [inline cmd/brimcap/ztests/analyze-process-error.yaml:2]: Maybe say `cmd: ./errorproc.sh` in the configuration files so you don't need this?

File: cmd/brimcap/ztests/analyze-process-error.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        - cmd: errorproc.sh
+  - name: config2.yaml
+    data: |
+      analyzers:
+        - cmd: errorproc.sh
+          stderr: stderr.out
+  - name: errorproc.sh
+    data: |
+      #!/bin/bash
+
```
Comment (nwt) [inline cmd/brimcap/ztests/analyze-process-error.yaml:26]:
```suggestion
```

**Review by nwt — APPROVED** [review]

**Comment by philrz** [comment]
I understand the spirit of what's being achieved here and the test makes sense. However, it's unclear to me what's meant by an "analyzer process" beyond the kind of single script as it's being used in the test. I'd guess that this is what would allow a user to do things like point at alternate Zeek/Suricata or add additional processes in the pipeline beyond the current combo of pcap index + Zeek/Suricata + reading the logs via the Zed platform, but I can't tell from the test example if that's already possible, and if so, how. My primary motive for asking is that I'd like to take the verification a step further by doing something like pointing at an intentionally broken Zeek/Suricata and seeing how this compares to the prior debug experience.

File: analyzer/process.go (hunk 2 above)
Comment (mattnibs) [reply analyzer/process.go:114]: Idk, then you have to do pointers every time you pass it into fmt.Fprint, why do you think this is better?

File: analyzer/config.go (hunk 1 above)
Comment (mattnibs) [reply analyzer/config.go:18]: Idk, don't you think stdout without path is easier to understand than `stdout_path`?

File: cmd/brimcap/ztests/analyze-process-error.yaml (hunk 7 above)
Comment (mattnibs) [reply cmd/brimcap/ztests/analyze-process-error.yaml:2]: Does not work, im thinking because cwd for the process is a tempdir?

**Comment by mattnibs** [comment]
> However, it's unclear to me what's meant by an "analyzer process" beyond the kind of single script as it's being used in the test.

@philrz all brimcap analyze really does is:
1. Execute a list of processes (creates a tempdir for each process)
2. Tee a pcap file into stdin for each process
3. Tail a the log files that are emitted

I've called them analyzers because they're really just performing analysis on a stream of pcaps. The generalization helps make it more flexible over just hardcoding in zeek and suricata. Indeed this is helpful for writing tests (which is important) but I believe there are other good things about this approach. For starters it makes the code simpler, it's just configs based on the three steps above, no specialized logic like we have in zqd. Once we get rid of the runners I imagine it would be useful to users to see the default config, for instance how we run zeek by basically feeding it `zeek -C -r -`, and see that it's easy for them to customize themselves. Additionally you could add analyzer processes, though I'm not sure if other ones exist, but for instance for brimcap index is an analyzer process in its own right, we could have this as part of the configuration.

**Pull request merged**